### PR TITLE
Replace experimental time API in TraceEncoder with custom implementation of TimeMark

### PR DESCRIPTION
**Pull request by JoelBryceAnderson**
* Replace experimental time API in TraceEncoder with custom implementation of TimeMark
* The default implementation, TraceEncoderTimeMark, holds a starting value from `System.nanoTime()` which it compares to future calls to `System.nanoTime()` to calculate elapsed time

### CODE REVIEW

File: trace-encoder/src/main/java/com/squareup/tracing/TimeMark.kt
Comment (rjrjr) [inline trace-encoder/src/main/java/com/squareup/tracing/TimeMark.kt:3]: Needs kdoc.

File: trace-encoder/src/main/java/com/squareup/tracing/TraceEncoder.kt
Comment (rjrjr) [inline trace-encoder/src/main/java/com/squareup/tracing/TraceEncoder.kt:22]: It's still a `TimeMark`, no?

```suggestion
 * @param start The [TimeMark] to consider the beginning timestamp of the trace. All
```

**Comment by rjrjr** [comment]
Run `./gradlew trace-encoder:apiCheck` to fix the API check.

File: trace-encoder/src/main/java/com/squareup/tracing/TraceEncoder.kt
Comment (JoelBryceAnderson) [reply trace-encoder/src/main/java/com/squareup/tracing/TraceEncoder.kt:22]: Tried to revert this change and I think I ended up dropping the commit. Will fix.

**Comment by zach-klippenstein** [comment]
I'm not sure I understand what this is actually fixing. I just ran the dungeon sample on `main` and everything worked fine. Could you give a little more context?

**Comment by JoelBryceAnderson** [comment]
> I'm not sure I understand what this is actually fixing. I just ran the dungeon sample on `main` and everything worked fine. Could you give a little more context?

To be honest, I'm not entirely sure of the root cause, but there seems to be an issue invoking the method `TimeMark#elapsedNow` inside `TraceEncoder` from applications using kotlin 1.5.0. I'm curious if it's some kind of version miss-match? Essentially the experimental time API is't behaving properly and this PR works around that.

**Comment by rjrjr** [comment]
The experimental warning on the Time classes particularly stress that it's a bad idea for libraries to use those methods, as binary compatibility is not assured.

I think what @JoelBryceAnderson found out is that our 1.5-based app is puking when it causes that call to be made against a 1.4-based release of the library.

**Review by rjrjr — APPROVED** [review]
LGTM, but let's see if Zach buys my library / app kotlin version mismatch theory before merging.

**Comment by zach-klippenstein** [comment]
Ah, I forgot POS was still on such an old release. That makes lots of sense, and yea probably shouldn't use experimental stuff.

**Review by zach-klippenstein — APPROVED** [review]

**Pull request merged**
